File: sdk/src/main/java/net/faithgen/sdk/http/ServerResponseListener.java
package net.faithgen.sdk.http;

import com.android.volley.VolleyError;

import net.faithgen.sdk.singletons.GSONSingleton;
import net.faithgen.sdk.utils.Constants;

import java.io.UnsupportedEncodingException;

public interface ServerResponseListener {
    void onResponse(String serverResponse);

    void onError(ErrorResponse errorResponse);

    public static ErrorResponse makeErrorResponse(VolleyError error) {
        ErrorResponse errorResponse = null;
        if (error.networkResponse != null) {
            String responseBody = null;
            try {
                responseBody = new String(error.networkResponse.data, "utf-8");
                errorResponse = GSONSingleton.getInstance().getGson().fromJson(responseBody, ErrorResponse.class);
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
                return null;
            }
        } else {
            errorResponse = new ErrorResponse();
            errorResponse.setCode("SER");
            errorResponse.setStatus(500);
            errorResponse.setMessage(Constants.SERVER_ERROR);
        }
        return errorResponse;
    }
}

File: sdk/src/main/java/net/faithgen/sdk/interfaces/ServerResponseListener.kt
package net.faithgen.sdk.interfaces

import net.faithgen.sdk.http.ErrorResponse

/**
 * Listens to HTTP events
 */
interface ServerResponseListener {
    /**
     * Handles the response on a successful request
     *
     * @param serverResponse The response from the server
     */
    fun onResponse(serverResponse: String?)

    /**
     * Handles the response on a failed request
     *
     * @param errorResponse The manufactured error response from Volley
     */
    fun onError(errorResponse: ErrorResponse?)
}